packages/structures/src/channels/Channel.ts

```diff
@@ -0,0 +1,205 @@
+import { DiscordSnowflake } from '@sapphire/snowflake';
+import type {
+	TextChannelType,
+	ThreadChannelType,
+	APIChannel,
+	APIPartialChannel,
+	ChannelType,
+	GuildChannelType,
+	GuildTextChannelType,
+	ChannelFlags,
+	APIThreadChannel,
+} from 'discord-api-types/v10';
+import { Structure } from '../Structure.js';
+import { ChannelFlagsBitField } from '../bitfields/ChannelFlagsBitField.js';
+import { kData } from '../utils/symbols.js';
+import type { ChannelPermissionMixin } from './mixins/ChannelPermissionMixin.js';
+import type { ChannelWebhookMixin } from './mixins/ChannelWebhookMixin.js';
+import type { DMChannelMixin } from './mixins/DMChannelMixin.js';
+import type { GuildChannelMixin } from './mixins/GuildChannelMixin.js';
+import type { TextChannelMixin } from './mixins/TextChannelMixin.js';
+import type { ThreadChannelMixin } from './mixins/ThreadChannelMixin.js';
+import type { ThreadOnlyChannelMixin } from './mixins/ThreadOnlyChannelMixin.js';
+import type { VoiceChannelMixin } from './mixins/VoiceChannelMixin.js';
+
+export type PartialChannel = Channel<ChannelType, Exclude<keyof APIChannel, keyof APIPartialChannel>>;
+
+/**
+ * @internal
+ */
+export type ChannelDataType<Type extends ChannelType> = Type extends
+	| ChannelType.AnnouncementThread
+	| ChannelType.PrivateThread
+	| ChannelType.PublicThread
+	? APIThreadChannel<Type>
+	: Extract<APIChannel, { type: Type }>; // TODO: this still breaks without the (new) special handling, find out why
```

Comment:
Confused about the todo and why we don't just do `APIChannel & { type: Type }`

Reply:
I'm confused too, hence the "find out why". But I assume it is because `APIPublicThreadChannel` etc. are types and not interfaces. Will try something and maybe PR another dtypes if that's the issue.